src/dataflow/src/render/join.rs

```diff
@@ -46,46 +50,90 @@ where
+                        demand_projection.push(column);
+                    } else {
+                        demand_projection.push(output_arity + dummies.len());
+                        dummies.push(ScalarExpr::literal_ok(Datum::Dummy, typ));
+                    }
+                }
+                (dummies, demand_projection)
+            } else {
+                (Vec::new(), (0..output_arity).collect::<Vec<_>>())
+            };
 
+            let (map, filter, project) = map_filter_project.as_map_filter_project();
+
+            let map_filter_project = MapFilterProject::new(output_arity)
+                .map(dummies)
+                .project(demand_projection)
+                .map(map)
+                .filter(filter)
+                .project(project);
+
+            // Other than the stream of updates, our loop-carried state are these
+            // three variables: `column_map`, `equivalences`, and `mfp`, which
+            // record the current locations of extended output columns, and what
+            // work remains to be done on them (other than the joining itself).
+            let mut column_map = HashMap::new();
+            for column in input_mapper.global_columns(*start) {
+                column_map.insert(column, column_map.len());
+            }
             // We maintain a private copy of `equivalences`, which we will digest
             // as we produce the join.
             let mut equivalences = equivalences.clone();
             for equivalence in equivalences.iter_mut() {
                 equivalence.sort();
                 equivalence.dedup();
             }
-
-            let input_mapper = expr::JoinInputMapper::new(inputs);
-
-            // Unwrap demand
-            // TODO: If we pushed predicates into the operator, we could have a
-            // more accurate view of demand that does not include the support of
-            // all predicates.
-            let demand = demand.clone().unwrap_or_else(|| (0..arity).collect());
+            // We maintain a private copy of `map_filter_project`, which we will
+            // digest as we produce the join.
+            let mut mfp = map_filter_project;
 
             // This collection will evolve as we join in more inputs.
+            // TODO(mcsherry): this clashes with the subsequent arrangement-based input.
+            // TODO(mcsherry): determine and apply closure here in `flat_map_ref` form.
+            // TODO(mcsherry): If we plan to use an arrangement, should one exist, then
+            // this is wasteful as it instantiates all rows which are then dropped.
             let (mut joined, mut errs) = self.collection(&inputs[*start]).unwrap();
 
-            // Maintain sources of each in-progress column.
-            let mut source_columns = input_mapper.global_columns(*start).collect::<Vec<_>>();
+            // NOTE(mcsherry): ideally this code is rarely/never relevant, as the associated logic
```

Comment:
this seems important? i can't tell if this should be resolved before merging but if not maybe it deserves to be its own issue?

Reply:
I think it is fine? The intent is that in the fullness of time, operators on this input probably get pushed further down, especially when there is not an arrangement. But, we don't really know what is up here, at the moment (like, just in the context of pushing in to join).

Reply:
ok just wanted to double check

Reply:
I moved the comment inside the conditional, to make it clearer that it was the "linear operator, no arrangement" case that we would ideally excise. I think it is a bit like the application at the end of the join chain though: ideally it would be absent, but there are probably mysterious legitimate reasons why it cannot be.